base/array.jl

```diff
@@ -1805,15 +1805,19 @@ julia> filter(isodd, a)
 """
 filter(f, As::AbstractArray) = As[map(f, As)::AbstractArray{Bool}]
 
-function filter!(f, a::Vector)
-    insrt = 1
+function filter!(f, a::AbstractVector)
+    idx = eachindex(a)
+    state = start(idx)
+    (i, state) = next(idx, state)
```

Comment:
Sorry, I had missed this: you need a `done(idx, state) && return a` check before this for empty arrays. The tests didn't catch this since the implementation for `Array` doesn't throw an error (and just returns `state+1`), but in general it could fail.

BTW, you can remove parentheses on LHS for tuple assignment (here and below).